bin/stPreprocess.py

```diff
@@ -0,0 +1,124 @@
+#!/opt/conda/bin/python
+
+import sys
+import os
+import scanpy as sc
+import numpy as np
+import pandas as pd
+from matplotlib import pyplot as plt
+import scipy.stats
+from scipy.sparse import csr_matrix
+
+# See more settings at:
+# https://scanpy.readthedocs.io/en/stable/generated/scanpy._settings.ScanpyConfig.html
+sc.settings.figdir = sys.argv[1]
+
+if not os.path.exists(sys.argv[1] + 'show/'):
+    os.makedirs(sys.argv[1] + 'show/')
+
+f_temp = np.load(sys.argv[1] + '/' + sys.argv[2] + '_factors.npz')
+f_temp = f_temp[list(f_temp.keys())[0]]
+print(f_temp.shape)
+
+st_adata = sc.read(sys.argv[1] + '/' + sys.argv[3])
+print(st_adata.shape)
+
+st_adata.obs['norm_factors'] = pd.Series(index=st_adata.obs[st_adata.obs['in_tissue']==1].index, data=f_temp).reindex(st_adata.obs.index)
+
+mito = pd.read_csv(sys.argv[4], index_col=['Symbol', 'MCARTA2_LIST'], delimiter='\t')['EnsemblGeneID']
+mito = mito.xs(1, level='MCARTA2_LIST').sort_index().reset_index()
+print(mito)
+
+st_adata.var["mt"] = st_adata.var_names.isin(mito['Symbol'])
+sc.pp.calculate_qc_metrics(st_adata, qc_vars=["mt"], inplace=True)
+
+plt.rcParams["figure.figsize"] = (6, 6)
+
+keys = ["in_tissue", "pct_counts_mt", "total_counts", "n_genes_by_counts"]
+st_adata_in = st_adata[st_adata.obs['in_tissue']==1].copy()
+sc.pl.spatial(st_adata_in, img_key="hires", color=keys, save='/st_QC_in.png')
+
+keys = ["pct_counts_mt", "total_counts", "n_genes_by_counts"]
+st_adata_out = st_adata[st_adata.obs['in_tissue']!=1].copy()
+sc.pp.filter_cells(st_adata_out, min_counts=500)
+sc.pp.filter_cells(st_adata_out, min_genes=250)
+sc.pp.filter_genes(st_adata_out, min_cells=1)

```

Comment:
This filtering doesn't seem to be used for anything other than the QC plot?

Reply:
Here it filters `st_adata_out` for QC only, and below the `st_adata` for all downstream parts, the flow needs to be optimized to avoid redundant code